cpp/src/arrow/compute/kernels/scalar_hash_test.cc

```diff
@@ -0,0 +1,336 @@
+      c_type actual_hash = res_array->GetValues<c_type>(1)[val_ndx];
+      if (arr->IsNull(val_ndx)) {
+        ASSERT_EQ(0, actual_hash);
+      } else {
+        ASSERT_EQ(exp[val_ndx], actual_hash);
+      }
+    }
+  }
+
+  template <typename c_type>
+  std::vector<c_type> HashPrimitive(const std::shared_ptr<Array>& arr) {
+    std::vector<c_type> hashes(arr->length());
+    // Choose the Hasher type conditionally based on c_type
+
+    if constexpr (std::is_same<c_type, uint64_t>::value) {
+      Hashing64::HashFixed(false, static_cast<uint32_t>(arr->length()),
+                           arr->type()->bit_width() / 8,
+                           arr->data()->GetValues<uint8_t>(1), hashes.data());
+    } else {
+      Hashing32::HashFixed(::arrow::internal::CpuInfo::GetInstance()->hardware_flags(),
+                           false, static_cast<uint32_t>(arr->length()),
+                           arr->type()->bit_width() / 8,
+                           arr->data()->GetValues<uint8_t>(1), hashes.data(), nullptr);
+    }
+
+    return hashes;
+  }
+
+  template <typename c_type>
+  std::vector<c_type> HashBinaryLike(const std::shared_ptr<Array>& arr) {
+    std::vector<c_type> hashes(arr->length());
+    auto length = static_cast<uint32_t>(arr->length());
+    auto values = arr->data()->GetValues<uint8_t>(2);
+    if constexpr (std::is_same<c_type, uint64_t>::value) {
+      if (arr->type_id() == Type::LARGE_BINARY || arr->type_id() == Type::LARGE_STRING) {
+        Hashing64::HashVarLen(false, length, arr->data()->GetValues<uint64_t>(1), values,
+                              hashes.data());
+      } else {
+        Hashing64::HashVarLen(false, length, arr->data()->GetValues<uint32_t>(1), values,
+                              hashes.data());
+      }
+    } else {
+      auto hw_flags = ::arrow::internal::CpuInfo::GetInstance()->hardware_flags();
+      if (arr->type_id() == Type::LARGE_BINARY || arr->type_id() == Type::LARGE_STRING) {
+        Hashing32::HashVarLen(hw_flags, false, length,
+                              arr->data()->GetValues<uint64_t>(1), values, hashes.data(),
+                              nullptr);
+      } else {
+        Hashing32::HashVarLen(hw_flags, false, length,
+                              arr->data()->GetValues<uint32_t>(1), values, hashes.data(),
+                              nullptr);
+      }
+    }
+    return hashes;
+  }
+
+  void CheckDeterminisic(const std::string& func, const std::shared_ptr<Array>& arr) {
+    // Check that the hash is deterministic
+    ASSERT_OK_AND_ASSIGN(Datum res1, CallFunction(func, {arr}));
+    ASSERT_OK_AND_ASSIGN(Datum res2, CallFunction(func, {arr}));
+    ASSERT_EQ(res1.length(), arr->length());
+    if (func == "hash64") {
+      ASSERT_EQ(res1.type()->id(), Type::UINT64);
+    } else if (func == "hash32") {
+      ASSERT_EQ(res1.type()->id(), Type::UINT32);
+    } else {
+      FAIL() << "Unknown function: " << func;
+    }
+    AssertDatumsEqual(res1, res2, true);
```

Comment:
Ok, but should also check that slicing doesn't change the output.

Reply:
What do you mean? Executing the kernel on a slice?

Reply:
Check that `hash(x[m:n]) == hash(x)[m:n]`

Reply:
It has been failing, added test and fixed the implementation.